sklearn/decomposition/nmf.py

```diff
@@ -78,6 +78,24 @@ def _safe_compute_error(X, W, H):
     return error
 
 
+def _compute_regularization(alpha, l1_ratio, regularization, n_samples,
+                            n_features):
+    """Compute L1 and L2 regularization coefficients for W and H"""
+    alpha_H = 0.
+    alpha_W = 0.
+    # The priors for W and H are scaled differently.
+    if regularization in ('both', 'components'):
+        alpha_H = float(alpha) * n_samples
```

Comment:
Here is the proper scaling of `alpha`.
All other modifications in this PR are just a clean way to pass the same regularization to every solvers.